cas-server-webapp-actions/src/main/java/org/jasig/cas/web/flow/GenerateServiceTicketAction.java

```diff
@@ -39,19 +41,37 @@
     private CentralAuthenticationService centralAuthenticationService;
 
     @NotNull
-    @Autowired(required=false)
+    @Autowired
     @Qualifier("defaultAuthenticationSystemSupport")
     private AuthenticationSystemSupport authenticationSystemSupport = new DefaultAuthenticationSystemSupport();
 
+    @Autowired
+    @Qualifier("defaultAuthenticationSupport")
+    private TicketRegistrySupport ticketRegistrySupport;
+
     @Override
     protected Event doExecute(final RequestContext context) {
         final Service service = WebUtils.getService(context);
         final String ticketGrantingTicket = WebUtils.getTicketGrantingTicketId(context);
 
         try {
-            final Credential credential = WebUtils.getCredential(context);
-            final AuthenticationResult authenticationResult =
-                    this.authenticationSystemSupport.handleAndFinalizeSingleAuthenticationTransaction(service, credential);
+            /**
+             * In the initial primary authentication flow, credentials are cached and available.
+             * Since they are authenticated as part of submission first, there is no need to doubly
+             * authenticate and verify credentials.
+             *
+             * In subsequent authentication flows where a TGT is available and only an ST needs to be
+             * created, there are no cached copies of the credential, since we do have a TGT available.
+             * So we will simply grab the available authentication and produce the final result based on that.
+             */
+            final Authentication authentication = ticketRegistrySupport.getAuthenticationFrom(ticketGrantingTicket);
+            if (authentication == null) {
+                throw new InvalidTicketException(
+                        new AuthenticationException("No authentication found for ticket " + ticketGrantingTicket), ticketGrantingTicket);
+            }
+            final AuthenticationResultBuilder authenticationResultBuilder = authenticationSystemSupport
```

Comment:
I started with that solution, but then realized that since we already have the TGT cached, I should be able to easily establish an authN context/result from that without getting SWF involved.  The approach you have in mind is implemented in MFA where we need that initial context for subsequent flows, so I decided to reuse the existing APIs for now, and provide support for SWF caching when MFA merges in. (The MFA branch does cache the result of course because flows need it later)